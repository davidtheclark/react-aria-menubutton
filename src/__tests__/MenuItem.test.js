const React = require('react');
const ReactDOMServer = require('react-dom/server');
const shallow = require('enzyme').shallow;
const shallowToJson = require('enzyme-to-json').shallowToJson;
const ManagerContext = require('../ManagerContext');
const MenuItem = require('../MenuItem');
const MockWrapper = require('./helpers/MockWrapper');
const createMockKeyEvent = require('./helpers/createMockKeyEvent');
const createMockManager = require('./helpers/createMockManager');

const el = React.createElement;

describe('<MenuItem>', function () {
  let shallowOptions;
  let ambManager;

  beforeEach(function () {
    ambManager = createMockManager();
    shallowOptions = {
      wrappingComponent: ManagerContext.Provider,
      wrappingComponentProps: { value: ambManager },
    };
  });

  it('DOM with only required props', function () {
    const menuItem = el(MenuItem, null, 'foo');
    const wrapper = shallow(menuItem, shallowOptions).dive().dive();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('DOM with all possible props and element child', function () {
    const menuItem = el(
      MenuItem,
      {
        className: 'foobar',
        id: 'hogwash',
        tag: 'li',
        style: { right: '1em' },
        text: 'horse',
        value: 'lamb',
        'data-something-something': 'seven', // arbitrary prop
      },
      el('a', { href: '#' }, 'foo')
    );
    const wrapper = shallow(menuItem, shallowOptions).dive().dive();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('click without specified value prop', function () {
    const mockEvent = { bee: 'baa' };
    const menuItem = el(MenuItem, null, 'foo');
    const wrapper = shallow(menuItem, shallowOptions).dive().dive();
    wrapper.simulate('click', mockEvent);
    expect(ambManager.handleSelection).toHaveBeenCalledTimes(1);
    expect(ambManager.handleSelection.mock.calls[0]).toEqual([
      'foo',
      mockEvent,
    ]);
  });

  it('click with specified value prop', function () {
    const mockEvent = { bee: 'baa' };
    const menuItem = el(MenuItem, { value: 'bar' }, 'foo');
    const wrapper = shallow(menuItem, shallowOptions).dive().dive();
    wrapper.simulate('click', mockEvent);
    expect(ambManager.handleSelection).toHaveBeenCalledTimes(1);
    expect(ambManager.handleSelection.mock.calls[0]).toEqual([
      'bar',
      mockEvent,
    ]);
  });

  it('click with specified value prop', function () {
    const mockEnterEvent = createMockKeyEvent('Enter');
    const mockSpaceEvent = createMockKeyEvent(' ');
    const mockEscapeEvent = createMockKeyEvent('Escape');
    const menuItem = el(MenuItem, null, 'foo');
    const wrapper = shallow(menuItem, shallowOptions).dive().dive();

    wrapper.simulate('keyDown', mockEnterEvent);
    wrapper.simulate('keyDown', mockSpaceEvent);
    wrapper.simulate('keyDown', mockEscapeEvent);
    expect(ambManager.handleSelection).toHaveBeenCalledTimes(2);
    expect(ambManager.handleSelection.mock.calls[0]).toEqual([
      'foo',
      mockEnterEvent,
    ]);
    expect(ambManager.handleSelection.mock.calls[1]).toEqual([
      'foo',
      mockSpaceEvent,
    ]);
  });
});

describe('<MenuItem> rendered via renderToString', function () {
  it('does not throw', function () {
    const output = ReactDOMServer.renderToString(
      el(
        MockWrapper,
        { mockManager: createMockManager() },
        el(MenuItem, null, 'foo')
      )
    );
    expect(output).toMatchSnapshot();
  });
});
