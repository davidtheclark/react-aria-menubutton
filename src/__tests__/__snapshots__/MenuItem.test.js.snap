// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuItem> DOM with all possible props and element child 1`] = `
<li
  className="foobar"
  data-something-something="seven"
  id="hogwash"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="menuitem"
  style={
    {
      "right": "1em",
    }
  }
  tabIndex="-1"
>
  <a
    href="#"
  >
    foo
  </a>
</li>
`;

exports[`<MenuItem> DOM with only required props 1`] = `
<div
  onClick={[Function]}
  onKeyDown={[Function]}
  role="menuitem"
  tabIndex="-1"
>
  foo
</div>
`;

exports[`<MenuItem> rendered via renderToString does not throw 1`] = `"<div><div role="menuitem" tabindex="-1">foo</div></div>"`;
